File: app/src/main/java/com/devpilot/weatherkok/who/WhoActivity.java
package com.devpilot.weatherkok.who;

import androidx.appcompat.app.AppCompatActivity;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.provider.ContactsContract;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.Toast;

import com.devpilot.weatherkok.R;
import com.devpilot.weatherkok.who.kakao.kotlin.KakaoApplication;
import com.devpilot.weatherkok.who.kakao.kotlin.SendKakao;
import com.facebook.share.model.ShareLinkContent;
import com.facebook.share.widget.ShareDialog;

public class WhoActivity extends AppCompatActivity {

    ImageView mIvKakaoTalk;
    ImageView mIvFacebook;
    ImageView mIvLine;
    String name;
    String number;
    Context mContext;
    KakaoApplication mKakaoApplication;
    SendKakao mSendKakao;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_who);
        mContext = getApplicationContext();

//        mIvKakaoTalk = findViewById(R.id.who_kakao);
//        mIvLine = findViewById(R.id.who_line);

        //페이스북 공유
//        mIvFacebook.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                ShareLinkContent content = new ShareLinkContent.Builder()
//                        .setContentTitle("페이스북 공유 링크입니다.")
//                        .setImageUrl(Uri.parse("게시물에 표시될 썸네일 이미지의 url"))
//                        .setContentUrl(Uri.parse("공유될 링크"))
//                        .setContentDescription("2~4개의 문장으로 구성된 콘텐츠 설명")
//                        .build();
//                ShareDialog shareDialog = new ShareDialog((Activity) getApplicationContext());
//            }
//        });


        //라인 메세지 보내는걸로?할까?
//        mIvLine.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//
//                Intent startLink = getPackageManager().getLaunchIntentForPackage("jp.naver.line.android");
//                String url;
//                String message = "공유하기";
//                    if (startLink != null) {
//                        String encodeMessage = URLEncoder.encode(message, "UTF-8");
//                        url = "http://line.me/R/msg/text/" + encodeMessage;
//                    } else {
//                        url = "https://play.google.com/store/apps/details?id=jp.naver.line.android";
//                    }
//                Intent intent = new Intent(Intent.ACTION_VIEW);
//                intent.setData(Uri.parse(url));
//                startActivity(intent);
//            }
//        });

        mIvKakaoTalk.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mSendKakao = new SendKakao();
                mSendKakao.sendKakaoMessage(mContext);

                //Intent intent = new Intent(mContext, SendKakao.class);
                //mContext.startActivity(intent);
            }
        });

    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        finish();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {

        super.onActivityResult(requestCode, resultCode, data);
    }

}

File: app/src/main/java/com/devpilot/weatherkok/who/kakao/kotlin/WhoActivity.kt
package com.devpilot.weatherkok.who.kakao.kotlin

import android.content.*
import android.net.Uri
import android.os.Bundle
import android.provider.ContactsContract
import android.util.Log
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.devpilot.weatherkok.R
import com.kakao.sdk.common.util.KakaoCustomTabsClient
import com.kakao.sdk.link.LinkClient
import com.kakao.sdk.link.WebSharerClient
import com.kakao.sdk.template.model.Link
import com.kakao.sdk.template.model.TextTemplate
import java.io.UnsupportedEncodingException
import java.net.URLEncoder

class WhoActivity : AppCompatActivity() {
    val TAG : String = "WhoActivity_kotlin"
    var name: String? = null
    var number: String? = null
    var mKakaoApplication: KakaoApplication? = null
    var mSendKakao: SendKakao? = null
    var messageText: String? = ""

    @Override
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_who)
        var mContext: Context? = null

        mContext = applicationContext
        var mIvKakaoTalk: ImageView? = null
        var LlRegisterTotal: LinearLayout? = null
        var mIvLine: ImageView? = null
        var mIvUrl: ImageView? = null

        //mIvKakaoTalk = findViewById<ImageView>(R.id.who_kakao)

        LlRegisterTotal = findViewById(R.id.ll_register_total)
        //mIvLine = findViewById<ImageView>(R.id.who_line)
        mIvUrl = findViewById<ImageView>(R.id.who_url)

        val intent: Intent = getIntent()
        var where: String = ""
        var whenString: String = ""

        if(intent!=null){
            where = intent.getStringExtra("where").toString();
            whenString = intent.getStringExtra("when").toString()
        }

        messageText = getString(R.string.share_message) + packageName

//        mIvKakaoTalk.setOnClickListener(View.OnClickListener {
//            sendKakaoMessage(mContext, messageText!!)
//        })
//
//        mIvLine.setOnClickListener(View.OnClickListener {
//
//            postAsLine(messageText)
//
//        })

        LlRegisterTotal.setOnClickListener(View.OnClickListener {

            onBackPressed()

        })

        mIvUrl.setOnClickListener(View.OnClickListener {

            var clipboardManager = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            var clipData = ClipData.newPlainText("URL", getString(R.string.play_store_url) + packageName)
            clipboardManager.setPrimaryClip(clipData)

            Toast.makeText(mContext, "URL이 복사되었습니다.", Toast.LENGTH_SHORT).show();
        })

    }

    @Throws(UnsupportedEncodingException::class)
    fun postAsLine(message: String?) {
        val startLink: Intent? = packageManager.getLaunchIntentForPackage("jp.naver.line.android")
        val url: String

        url = if (startLink != null) {
            val encodedMessage = URLEncoder.encode(message, "UTF-8")
            "http://line.me/R/msg/text/$encodedMessage"
        } else {
            "https://play.google.com/store/apps/details?id=jp.naver.line.android"
        }
        val intent = Intent(Intent.ACTION_VIEW)
        intent.data = Uri.parse(url)
        startActivity(intent)
    }

    fun sendKakaoMessage(context : Context, messageText: String) {

        val defaultFeed = TextTemplate(
            text = messageText.trimIndent(),
            link = Link(
                webUrl = "https://developers.kakao.com",
                mobileWebUrl = "https://developers.kakao.com"
            )
        )

        // 피드 메시지 보내기

// 카카오톡 설치여부 확인
        if (LinkClient.instance.isKakaoLinkAvailable(context)) {
            // 카카오톡으로 카카오링크 공유 가능
            LinkClient.instance.defaultTemplate(context, defaultFeed) { linkResult, error ->
                if (error != null) {
                    Log.e(TAG, "카카오링크 보내기 실패", error)
                }
                else if (linkResult != null) {
                    Log.d(TAG, "카카오링크 보내기 성공 ${linkResult.intent}")
                    startActivity(linkResult.intent)

                    // 카카오링크 보내기에 성공했지만 아래 경고 메시지가 존재할 경우 일부 컨텐츠가 정상 동작하지 않을 수 있습니다.
                    Log.w(TAG, "Warning Msg: ${linkResult.warningMsg}")
                    Log.w(TAG, "Argument Msg: ${linkResult.argumentMsg}")
                }
            }
        } else {
            // 카카오톡 미설치: 웹 공유 사용 권장
            // 웹 공유 예시 코드
            val sharerUrl = WebSharerClient.instance.defaultTemplateUri(defaultFeed)

            // CustomTabs으로 웹 브라우저 열기

            // 1. CustomTabs으로 Chrome 브라우저 열기
            try {
                KakaoCustomTabsClient.openWithDefault(context, sharerUrl)
            } catch(e: UnsupportedOperationException) {
                // Chrome 브라우저가 없을 때 예외처리
            }

            // 2. CustomTabs으로 디바이스 기본 브라우저 열기
            try {
                KakaoCustomTabsClient.open(context, sharerUrl)
            } catch (e: ActivityNotFoundException) {
                // 인터넷 브라우저가 없을 때 예외처리
            }
        }


    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {

        super.onActivityResult(requestCode, resultCode, data)
    }


    override fun onBackPressed() {
        super.onBackPressed()

        finish()

    }
}